File: 05-Algorithms/code/python/seb_alg/cycle.py
"""Mirrors `typescript/src/cycle.ts` exactly."""

from __future__ import annotations

from typing import List, Sequence

from .graph_common import build_adj_list


def has_cycle(n: int, edges: Sequence[Sequence[int]], directed: bool) -> bool:
    """Detects whether the n-vertex graph has a cycle.

    - directed=True: white/gray/black DFS coloring; a gray vertex reached
      again means a back-edge, i.e. a cycle.
    - directed=False: a back-edge to an already-visited non-parent vertex
      indicates a cycle; parallel edges to the parent also count (tracked
      via edge-occurrence count, not just vertex identity).

    O(V + E).
    """
    adj = build_adj_list(n, edges, directed)

    if directed:
        color = [0] * n  # 0=white,1=gray,2=black
        for start in range(n):
            if color[start] != 0:
                continue
            stack: List[List[int]] = [[start, 0]]
            color[start] = 1
            while stack:
                frame = stack[-1]
                u, i = frame[0], frame[1]
                neighbors = adj[u]
                if i < len(neighbors):
                    v = neighbors[i]
                    frame[1] += 1
                    if color[v] == 1:
                        return True
                    if color[v] == 0:
                        color[v] = 1
                        stack.append([v, 0])
                else:
                    color[u] = 2
                    stack.pop()
        return False

    visited = [False] * n
    parent = [-1] * n
    for start in range(n):
        if visited[start]:
            continue
        stack = [start]
        visited[start] = True
        while stack:
            u = stack.pop()
            for v in adj[u]:
                if not visited[v]:
                    visited[v] = True
                    parent[v] = u
                    stack.append(v)
                elif v != parent[u]:
                    return True
                else:
                    occurrences = sum(1 for x in adj[u] if x == v)
                    if occurrences > 1:
                        return True
    return False

File: 05-Algorithms/code/typescript/src/cycle.ts
import { buildAdjList, type Edge } from "./graphCommon.js";

/**
 * Detects whether n-vertex graph has a cycle.
 *
 * - `directed=false`: a back-edge to an already-visited vertex that is not
 *   the immediate parent indicates a cycle (a parallel edge -- two vertices
 *   joined twice -- also counts, since it forms a length-2 cycle; the parent
 *   tracking is by edge occurrence, not just vertex, to catch that case).
 * - `directed=true`: uses the classic white/gray/black DFS coloring; a gray
 *   (on-stack) vertex reached again means a back-edge, i.e. a cycle.
 *
 * O(V + E).
 */
export function hasCycle(n: number, edges: Edge[], directed: boolean): boolean {
  const adj = buildAdjList(n, edges, directed);

  if (directed) {
    const color = new Array(n).fill(0); // 0=white,1=gray,2=black
    for (let start = 0; start < n; start++) {
      if (color[start] !== 0) continue;
      const stack: { u: number; i: number }[] = [{ u: start, i: 0 }];
      color[start] = 1;
      while (stack.length > 0) {
        const frame = stack[stack.length - 1];
        const neighbors = adj[frame.u];
        if (frame.i < neighbors.length) {
          const v = neighbors[frame.i++];
          if (color[v] === 1) return true;
          if (color[v] === 0) {
            color[v] = 1;
            stack.push({ u: v, i: 0 });
          }
        } else {
          color[frame.u] = 2;
          stack.pop();
        }
      }
    }
    return false;
  }

  const visited = new Array(n).fill(false);
  const parent = new Array(n).fill(-1);
  for (let start = 0; start < n; start++) {
    if (visited[start]) continue;
    const stack: number[] = [start];
    visited[start] = true;
    while (stack.length > 0) {
      const u = stack.pop() as number;
      for (const v of adj[u]) {
        if (!visited[v]) {
          visited[v] = true;
          parent[v] = u;
          stack.push(v);
        } else if (v !== parent[u]) {
          return true;
        } else {
          // v === parent[u]: could still be a genuine cycle if there are
          // parallel edges between u and its parent; count occurrences.
          const occurrences = adj[u].filter((x) => x === v).length;
          if (occurrences > 1) return true;
        }
      }
    }
  }
  return false;
}
